src/aws/flb_aws_compress.c

```diff
@@ -0,0 +1,231 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_log.h>
+#include <mbedtls/base64.h>
+
+#include <fluent-bit/aws/flb_aws_compress.h>
+#include <fluent-bit/flb_gzip.h>
+
+#include <stdint.h>
+
+#ifdef FLB_HAVE_ARROW
+#include "compression/arrow/compress.h"
+#endif
+
+struct compression_option {
+    int compression_type;
+    char *compression_keyword;
+    int(*compress)(void *in_data, size_t in_len, void **out_data, size_t *out_len);
+};
+
+/*
+ * Library of compression options
+ * AWS plugins that support compression will have these options.
+ * Referenced function should return -1 on error and 0 on success.
+ */
+static const struct compression_option compression_options[] = {
+    /* FLB_AWS_COMPRESS_NONE which is 0 is reserved for array footer */
+    {
+        FLB_AWS_COMPRESS_GZIP,
+        "gzip",
+        &flb_gzip_compress
+    },
+#ifdef FLB_HAVE_ARROW
+    {
+        FLB_AWS_COMPRESS_ARROW,
+        "arrow",
+        &out_s3_compress_arrow
+    },
+#endif
+    { 0 }
+};
+
+int flb_aws_compression_get_type(const char *compression_keyword)
+{
+    int ret;
+    const struct compression_option *o;
+
+    o = compression_options;
+
+    while (o->compression_type != 0) {
+        ret = strcmp(o->compression_keyword, compression_keyword);
+        if (ret == 0) {
+            return o->compression_type;
+        }
+        ++o;
+    }
+
+    flb_error("[aws_compress] unknown compression type: %s", compression_keyword);
+    return -1;
+}
+
+int flb_aws_compression_compress(int compression_type, void *in_data, size_t in_len,
+                                void **out_data, size_t *out_len)
+{
+    const struct compression_option *o;
+
+    o = compression_options;
+
+    while (o->compression_type != 0) {
+        if (o->compression_type == compression_type) {
+            return o->compress(in_data, in_len, out_data, out_len);
+        }
+        ++o;
+    }
+    
+    flb_error("[aws_compress] invalid compression type: %i", compression_type);
+    flb_errno();
+    return -1;
+}
+
+int flb_aws_compression_b64_truncate_compress(int compression_type, size_t max_out_len,
+                                             void *in_data, size_t in_len,
+                                             void **out_data, size_t *out_len)
+{
+    static const void *truncation_suffix = "[Truncated...]";
+    static const size_t truncation_suffix_len = 14;
+    static const double truncation_reduction_percent = 90; /* % out of 100 */
```

Comment:
these seem like they should be constants at the top of this file or in the header file?

Reply:
Also, can you explain to me what `truncation_reduction_percent` means and why

Reply:
static const means that these constants wont be recreated for every function call. Placing these inside of the function restricts the scope to the function. You can see a similar pattern here: https://github.com/fluent/fluent-bit/blob/9ccc6f89171d6792f9605b77b4efd27bc28f472a/src/flb_utils.c#L601

Reply:
I see. What is the truncation reduction percent then?

Reply:
Just got back to this.

The truncation system works by taking the old uncompressed size and compares it's ratio to the new compressed size. The old uncompressed size is reduced to make the compressed size match the expected truncation length.

However this is only an estimate, and assumes that the whole uncompressed data is equally random at the start and end.

If the uncompressed data is not equally random this estimate does work perfectly. Take for example "sjbrjrhwkabthihihihihihihihih"

This will actually turn into something like "xxxxxxxxx""yy" where x is from the first half of the data and y is the last half. Truncating the log with the ratio method gives a best guess, but the "truncation reduction percent" adds some grace room to account for not equally distributed randomness in the data.

The penalty for overestimating the truncated length is another one or several iterations. Please see the unit test case which has half Lorem I spum text and half uniform characters to see when a couple iterations are still needed, even with this truncation padding.

Reply:
It also helps to prevent near infinite compression iterations in edge cases.
https://github.com/fluent/fluent-bit/blob/c4f9cff36346ee49e4469b128a49f6196a24624e/tests/internal/aws_compress.c#L169-L201

Reply:
As I asked in the golang version of this- is there a justification for why 90% and this overal algo are ideal or make some good tradeoff between efficiency and minimizing data loss?

Reply:
please see: https://quip-amazon.com/bbHPAwzI3Kwg/Compression-Truncation

Reply:
Discussed internally. Resolving.